File: vane-core/src/main/java/org/oddlama/vane/core/menu/HeadFilter.java
package org.oddlama.vane.core.menu;

import java.util.List;
import java.util.stream.Collectors;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.oddlama.vane.core.material.HeadMaterial;
import org.oddlama.vane.core.menu.Menu.ClickResult;
import org.oddlama.vane.core.module.Context;

public class HeadFilter implements Filter<HeadMaterial> {

    private String str = null;

    public HeadFilter() {}

    @Override
    public void openFilterSettings(
        final Context<?> context,
        final Player player,
        final String filterTitle,
        final Menu returnTo
    ) {
        MenuFactory.anvilStringInput(
            context,
            player,
                filterTitle,
            new ItemStack(Material.PAPER),
            "?",
            (p, menu, s) -> {
                menu.close(p);
                str = s.toLowerCase();
                returnTo.open(p);
                return ClickResult.SUCCESS;
            }
        ).open(player);
    }

    @Override
    public void reset() {
        str = null;
    }

    private boolean filterByCategories(final HeadMaterial material) {
        return material.category().toLowerCase().contains(str);
    }

    private boolean filterByTags(final HeadMaterial material) {
        for (final var tag : material.tags()) {
            if (tag.toLowerCase().contains(str)) {
                return true;
            }
        }

        return false;
    }

    private boolean filterByName(final HeadMaterial material) {
        return material.name().toLowerCase().contains(str);
    }

    @Override
    public List<HeadMaterial> filter(final List<HeadMaterial> things) {
        if (str == null) {
            return things;
        }

        return things
            .stream()
            .filter(t -> filterByCategories(t) || filterByTags(t) || filterByName(t))
            .collect(Collectors.toList());
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/menu/HeadFilter.kt
package org.oddlama.vane.core.menu

import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.oddlama.vane.core.material.HeadMaterial
import org.oddlama.vane.core.module.Context

/**
 * Filter implementation for head selector entries.
 */
class HeadFilter : Filter<HeadMaterial> {
    /**
     * Current lowercase filter string.
     */
    private var str: String? = null

    /**
     * Opens an anvil input for editing the head filter.
     */
    override fun openFilterSettings(
        context: Context<*>,
        player: Player,
        filterTitle: String,
        returnTo: Menu?
    ) {
        MenuFactory.anvilStringInput(
            context,
            player,
            filterTitle,
            ItemStack(Material.PAPER),
            "?"
        ) { t1, t2, t3 ->
            t2!!.close(t1!!)
            str = t3?.lowercase()
            returnTo?.open(t1)
            Menu.ClickResult.SUCCESS
        }.open(player)
    }

    /**
     * Clears the active filter.
     */
    override fun reset() {
        str = null
    }

    /**
     * Returns whether a head matches by category.
     */
    private fun filterByCategory(material: HeadMaterial): Boolean =
        material.category?.lowercase()?.contains(str ?: return false) == true

    /**
     * Returns whether a head matches by tags.
     */
    private fun filterByTags(material: HeadMaterial): Boolean {
        val filter = str ?: return false
        return material.tags.any { it?.lowercase()?.contains(filter) == true }
    }

    /**
     * Returns whether a head matches by display name.
     */
    private fun filterByName(material: HeadMaterial): Boolean =
        material.name?.lowercase()?.contains(str ?: return false) == true

    /**
     * Filters head materials by name, category, or tags.
     */
    override fun filter(things: MutableList<HeadMaterial?>): MutableList<HeadMaterial?> {
        if (str == null) return things
        return things.filterTo(mutableListOf()) { t ->
            t != null && (filterByCategory(t) || filterByTags(t) || filterByName(t))
        }
    }
}
